lib/auth/auth_with_roles.go

```diff
@@ -1257,16 +1248,41 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
 		roles = utils.Deduplicate(roles)
 	}
 
-	// Extract the user and role set for whom the certificate will be generated.
-	user, err := a.GetUser(req.Username, false)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
 	checker, err := services.FetchRoles(roles, a.authServer, traits)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
+	switch {
+	case a.hasBuiltinRole(string(teleport.RoleAdmin)):
+		// builtin admins can impersonate anyone
+		// this is required for local tctl commands to work
+	case req.Username == a.context.User.GetName():
+		// users can impersonate themselves
+	default:
+		// check if this user is allowed to impersonate other users
+		err = a.context.Checker.CheckImpersonate(a.context.User, user, checker)
+		if err != nil {
+			err := trace.AccessDenied("user %q has requested to generate certs for %q.", a.context.User.GetName(), roles)
+			log.Warning(err)
+			if err := a.authServer.emitter.EmitAuditEvent(a.CloseContext(), &events.UserLogin{
+				Metadata: events.Metadata{
+					Type: events.UserLoginEvent,
+					Code: events.UserLocalLoginFailureCode,
+				},
+				Method: events.LoginMethodClientCert,
+				Status: events.Status{
+					Success:     false,
+					Error:       trace.Unwrap(err).Error(),
+					UserMessage: err.Error(),
+				},
+			}); err != nil {
+				log.WithError(err).Warn("Failed to emit local login failure event.")
+			}
+			return nil, trace.Wrap(err)
```

Comment:
This will reveal a full `user %q has requested to generate ....` error message instead of generic `access denied` like in other places. Is this intentional?

Reply:
I thought that would not leak any specific information, because that's what client has supplied